lib/lfrfid/protocols/protocol_paradox.c

```diff
@@ -0,0 +1,211 @@
+typedef struct {
+    FSKDemod* fsk_demod;
+} ProtocolParadoxDecoder;
+
+typedef struct {
+    FSKOsc* fsk_osc;
+    uint8_t encoded_index;
+} ProtocolParadoxEncoder;
+
+typedef struct {
+    ProtocolParadoxDecoder decoder;
+    ProtocolParadoxEncoder encoder;
+    uint8_t encoded_data[PARADOX_ENCODED_DATA_SIZE];
+    uint8_t data[PARADOX_DECODED_DATA_SIZE];
+} ProtocolParadox;
+
+ProtocolParadox* protocol_paradox_alloc(void) {
+    ProtocolParadox* protocol = malloc(sizeof(ProtocolParadox));
+    protocol->decoder.fsk_demod = fsk_demod_alloc(MIN_TIME, 6, MAX_TIME, 5);
+    protocol->encoder.fsk_osc = fsk_osc_alloc(8, 10, 50);
+
+    return protocol;
+};
+
+void protocol_paradox_free(ProtocolParadox* protocol) {
+    fsk_demod_free(protocol->decoder.fsk_demod);
+    fsk_osc_free(protocol->encoder.fsk_osc);
+    free(protocol);
+};
+
+uint8_t* protocol_paradox_get_data(ProtocolParadox* protocol) {
+    return protocol->data;
+};
+
+void protocol_paradox_decoder_start(ProtocolParadox* protocol) {
+    memset(protocol->encoded_data, 0, PARADOX_ENCODED_DATA_SIZE);
+};
+
+static bool protocol_paradox_can_be_decoded(ProtocolParadox* protocol) {
+    bool result = false;
+
+    do {
+        // check preamble
+        if(protocol->encoded_data[0] != 0b00001111 ||
+           protocol->encoded_data[PARADOX_ENCODED_DATA_LAST] != 0b00001111)
+            break;
+
+        for(uint32_t i = PARADOX_PREAMBLE_LENGTH; i < 96; i += 2) {
+            if(bit_lib_get_bit(protocol->encoded_data, i) ==
+               bit_lib_get_bit(protocol->encoded_data, i + 1)) {
+                return false;
+            }
+        }
+
+        result = true;
+    } while(false);
+
+    return result;
+}
+
+static void protocol_paradox_decode(uint8_t* encoded_data, uint8_t* decoded_data) {
+    for(uint32_t i = PARADOX_PREAMBLE_LENGTH; i < 96; i += 2) {
+        if(bit_lib_get_bits(encoded_data, i, 2) == 0b01) {
+            bit_lib_push_bit(decoded_data, PARADOX_DECODED_DATA_SIZE, 0);
+        } else if(bit_lib_get_bits(encoded_data, i, 2) == 0b10) {
+            bit_lib_push_bit(decoded_data, PARADOX_DECODED_DATA_SIZE, 1);
+        }
+    }
+    bit_lib_push_bit(decoded_data, PARADOX_DECODED_DATA_SIZE, 0);
+    bit_lib_push_bit(decoded_data, PARADOX_DECODED_DATA_SIZE, 0);
+    bit_lib_push_bit(decoded_data, PARADOX_DECODED_DATA_SIZE, 0);
+    bit_lib_push_bit(decoded_data, PARADOX_DECODED_DATA_SIZE, 0);
+}
+
+bool protocol_paradox_decoder_feed(ProtocolParadox* protocol, bool level, uint32_t duration) {
+    bool value;
+    uint32_t count;
+    bool result = false;
+
+    fsk_demod_feed(protocol->decoder.fsk_demod, level, duration, &value, &count);
+    if(count > 0) {
+        for(size_t i = 0; i < count; i++) {
+            bit_lib_push_bit(protocol->encoded_data, PARADOX_ENCODED_DATA_SIZE, value);
+            if(protocol_paradox_can_be_decoded(protocol)) {
+                protocol_paradox_decode(protocol->encoded_data, protocol->data);
+
+                result = true;
```

Comment:
remove `result` -> `return true`?

Reply:
A single exit point from a function is highly desirable.